OBJC_INTEROP.md

```diff
@@ -0,0 +1,124 @@
+
+### Name translation
+
+Objective-C classes are imported into Kotlin with their original names.
+Protocols are imported as interfaces with `Protocol` name suffix,
+i.e. `@protocol Foo` -> `interface FooProtocol`.
+These classes and interfaces are placed into package [specified in build configuration](#usage)
+(`platform.*` packages for preconfigured system frameworks).
+
+Names of Kotlin classes and interfaces are prefixed when imported to Swift/Objective-C.
+The prefix is derived from the framework name.
+
+### Top-level functions and properties
+
+Top-level Kotlin functions and properties are accessible as members of a special class.
+Each Kotlin package is translated into such a class.
+
+### Errors and exceptions
+
+Kotlin has no concept of checked exceptions, all Kotlin exceptions are unchecked.
+Swift has only checked errors. So if Swift or Objective-C code calls Kotlin method
+which throws an exception to be handled, then Kotlin method should be marked
+with `@Throws` annotation. In this case all Kotlin exceptions
+(except for instances of `Error`, `RuntimeException` and subclasses) are translated to
+Swift error/`NSError`.
+
+Note that the opposite translation is not implemented yet:
+Swift/Objective-C error-throwing methods aren't imported to Kotlin as
+exception-throwing.
+
+### Category members
+
+Members of Objective-C categories and Swift extensions are imported to Kotlin
+as extensions. That's why these declarations can't be overridden in Kotlin.
+And extension initializers aren't available as Kotlin constructors.
+
+### Number
+
+`NSNumber` is not automatically translated to Kotlin primitive types
+when used as Swift/Objective-C parameter type or return value.
+
+### NSMutableString
+
+`NSMutableString` Objective-C class is not available from Kotlin.
+All instances of `NSMutableString` are copied when passed to Kotlin.
+
+### Mutable collections
+
+Every Kotlin `MutableSet` is `NSMutableSet`, however the opposite is not true.
+To pass an object for Kotlin `MutableSet`, one can create this kind of Kotlin collection
+explicitly by either creating it in Kotlin with e.g. `mutableSetOf()`,
+or using `${prefix}MutableSet` class in Swift/Objective-C, where `prefix`
+is the framework names prefix.
+
+The same holds for `MutableMap`.
+
+## Casting between mapped types
+
+When writing Kotlin code, an object may require to be converted from Kotlin type
+to equivalent Swift/Objective-C type (or vice versa). In this case plain old
+Kotlin cast can be used, e.g.
+```
+val nsArray = listOf(1, 2, 3) as NSArray
+val string = nsString as String
+val nsNumber = 42 as NSNumber
+```
+
+## Subclassing
+
+Kotlin classes and interfaces can be subclassed by Swift/Objective-C classes
+and protocols.
+Currently a class that adopts Kotlin protocol should inherit `NSObject`
+(either directly or indirectly). Note that all Kotlin classes do inherit `NSObject`.
+
+Swift/Objective-C classes can be subclassed with Kotlin `final` class.
```

Comment:
Please mention, that non-final inheritance and complex object hierarchies are not yet supported.